pkg/device42/paginate.go

```diff
@@ -0,0 +1,92 @@
+package device42
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// Paginator is an interface that provides a concurrent fan-out to retrieve multiple
+// response bodies from a paginated API
+type Paginator interface {
+	BatchPagedRequests(ctx context.Context) ([][]byte, error)
+}
+
+// Device42Paginator implements the Paginator interface for the Device42 paginated APIs
+// See: https://api.device42.com/#api-get-limits-and-offsets
+type Device42Paginator struct {
+	Client   *http.Client
+	Endpoint *url.URL
+	Limit    int
+}
+
+type pagedPayload struct {
+	TotalCount int `json:"total_count"`
+}
+
+// BatchPagedRequests performs a current fan-out to retrieve all responses from a Device42
+// paginated API endpoint
+func (p *Device42Paginator) BatchPagedRequests(ctx context.Context) ([][]byte, error) {
```

Comment:
suggestion:

This is great! I think you found a good solution for abstracting out the pagination logic from the request types.

In light of @bkrebsbach 's comment above, though, I think we should start without concurrency, and experiment with adding it later if performance proves to be a problem. Ordinarily, making sequential requests would absolutely be the wrong approach, but here I think we should default to less load on Device42. 

In either case, though, I would suggest one more layer of abstraction to isolate the logic around making paged requests from the strategy for making them - as in, concurrently or serially, etc. This way, if we add some sort of throttled concurrent logic later, we don't have to modify the paging logic, or copy-paste the `makePagedRequest` logic across `Paginator` implementations.

This mainly means promoting the `makePagedRequest` logic to its own interface. Here is a snippet of what this might look like:

```go
type PagedResponse struct {
	TotalCount int  `json:"total_count"`
	Limit int `json:"limit"`
	Offset int `json:"offset"`
	Body []byte
}
	
func (r *PagedResponse) UnmarshalJSON(data []byte) error {
	type Alias PagedResponse
	respAlias := &Alias{}
	err := json.Unmarshal(data, &respAlias)
	if err != nil {
		return err
	}
	r.TotalCount = respAlias.TotalCount
	r.Limit = respAlias.Limit
	r.Offset = respAlias.Offset
	r.Body = data
	return nil
}

// PageRequester encapsulates the logic for requesting individual pages from a paged REST API
type PageRequester interface {
	MakePagedRequest(ctx context.Context, offset int) (PagedResponse, error)
}

// Paginator encapsulates the logic for requesting all of the pages from a paged REST API, including
// any concurrency that may be employed
type Paginator interface {
	BatchPagedRequests(ctx context.Context) ([]PagedResponse, error)
}

type Device42PageRequester struct {
	Client   *http.Client
	Endpoint *url.URL
}

// ThrottledConcurrentPaginator requests a max of MaxRequests pages concurrently.
type ThrottledConcurrentPaginator struct {
	Pager PageRequester
	Limit    int
	Semaphore sync.Locker
	MaxRequests int	
}

// SerialPaginator requests each page sequentially.
type SerialPaginator struct {
	Pager PageRequester
	Limit int
}

func (p *SerialPaginator) BatchPagedRequests(ctx context.Context) ([]PagedResponse, error){
	initialResponse, err := p.Pager.MakePagedRequest(ctx, 0)
	if err != nil {
		return nil, err
	}
	responses := make([]PagedResponse, initialResponse.TotalCount)
	for offset := p.Limit; offset <= initialResponse.TotalCount - p.Limit; offset = offset + p.Limit {
		resp, err := p.Pager.MakePagedRequest(ctx, offset)
		if err != nil {
			return nil, err
		}
		responses = append(responses, resp)
	}
	return responses, nil
}
```